### PR TITLE
Support  Writing Arrow files

### LINKED ISSUE #8504: Support writing to Arrow files
### Is your feature request related to a problem or challenge?

We currently support reading Arrow files:

https://github.com/apache/arrow-datafusion/blob/95ba48bd2291dd5c303bdaf88cbb55c79d395930/datafusion/sqllogictest/test_files/arrow_files.slt#L1

However, we  do not support writing them:

```
❯ copy (values (1)) to '/tmp/data.arrow';
This feature is not implemented: Writer not implemented for this format
```

### Describe the solution you'd like

I would like to be able to write to arrow files using the `COPY` command and `EXTERNAL TABLE`s 

The idea would be to implement `create_writer_physical_plan`

https://github.com/apache/arrow-datafusion/blob/95ba48bd2291dd5c303bdaf88cbb55c79d395930/datafusion/core/src/datasource/file_format/arrow.rs#L51

Following the model of the CSV file format and
https://github.com/apache/arrow-datafusion/blob/95ba48bd2291dd5c303bdaf88cbb55c79d395930/datafusion/core/src/datasource/file_format/csv.rs#L262-L290

Then add tests in copy.slt and arrow_file.slt

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devinjdangelo**
## Which issue does this PR close?

Closes #8504 

## What changes are included in this PR?

Implements initial support for writing out arrow files via `COPY TO` and `INSERT INTO` for listing tables.

## Are these changes tested?

Adds new sqllogictests to cover.

## Are there any user-facing changes?

Writing arrow files is now possible

### CODE REVIEW

File: datafusion/sqllogictest/test_files/insert_to_external.slt [hunk 1]
```diff
@@ -76,6 +76,44 @@ select * from dictionary_encoded_parquet_partitioned order by (a);
...
+create_local_path 'true',
+insert_mode 'append_new_files',
+);
+
+query T
+select * from dictionary_encoded_arrow_test_readback;
+----
+b
+
+query error DataFusion error: Arrow error: Schema error: project index 1 out of bounds, max field 1
```
Comment (devinjdangelo) [inline datafusion/sqllogictest/test_files/insert_to_external.slt:114]: Reading a table of partitioned arrow files fails with this error. I believe the arrow files themselves are fine as shown in the test above that reads one of the partitions correctly.

Comment (devinjdangelo) [reply datafusion/sqllogictest/test_files/insert_to_external.slt:114]: Possibly similar cause as #7816 ?

File: datafusion/sqllogictest/test_files/copy.slt [hunk 2]
```diff
@@ -230,6 +230,62 @@ select * from validate_csv_with_options;
...
+STORED AS arrow
+LOCATION 'test_files/scratch/copy/table.arrow';
+
+query IT
+select * from validate_arrow_file;
+----
+1 Foo
+2 Bar
+
+# Copy from dict encoded values to single arrow file
```
Comment (devinjdangelo) [inline datafusion/sqllogictest/test_files/copy.slt:251]: I know @tustvold's main concern was dictionaries. I _think_ this test shows we are OK, but let me know if I am overlooking something.

File: Cargo.toml [hunk 3]
```diff
@@ -53,24 +36,26 @@ arrow = { version = "49.0.0", features = ["prettyprint"] }
...
 datafusion-common = { path = "datafusion/common", version = "34.0.0" }
+datafusion-execution = { path = "datafusion/execution", version = "34.0.0" }
 datafusion-expr = { path = "datafusion/expr", version = "34.0.0" }
-datafusion-sql = { path = "datafusion/sql", version = "34.0.0" }
 datafusion-optimizer = { path = "datafusion/optimizer", version = "34.0.0" }
 datafusion-physical-expr = { path = "datafusion/physical-expr", version = "34.0.0" }
 datafusion-physical-plan = { path = "datafusion/physical-plan", version = "34.0.0" }
-datafusion-execution = { path = "datafusion/execution", version = "34.0.0" }
 datafusion-proto = { path = "datafusion/proto", version = "34.0.0" }
+datafusion-sql = { path = "datafusion/sql", version = "34.0.0" }
```
Comment (viirya) [inline Cargo.toml:56]: Hmm, are these changes related? Looks like just moving lines around? Maybe you can revert unrelated change to keep diff smaller?

Comment (devinjdangelo) [reply Cargo.toml:56]: I added arrow-ipc as a dependency for core, and ran cargo tomlfmt. I'm not sure why cargo tomlfmt changed so much of the formatting 🤔

File: datafusion/sqllogictest/test_files/insert_to_external.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/insert_to_external.slt:114]:
```suggestion
# https://github.com/apache/arrow-datafusion/issues/7816
query error DataFusion error: Arrow error: Schema error: project index 1 out of bounds, max field 1
```

File: datafusion/core/src/datasource/file_format/arrow.rs [hunk 4]
```diff
@@ -97,11 +110,191 @@ impl FileFormat for ArrowFormat {
...
+    }
+
+    async fn write_all(
+        &self,
+        data: SendableRecordBatchStream,
+        context: &Arc<TaskContext>,
+    ) -> Result<u64> {
+        // No props are supported yet, but can be by updating FileTypeWriterOptions
+        // to populate this struct and use those options to initialize the arrow_ipc::writer::FileWriter
+        let _arrow_props = self.config.file_type_writer_options.try_into_arrow()?;
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/arrow.rs:211]: I think we should track this as a follow on ticket and ideally leave a comment in the code pointing to the ticket so it eventually gets cleaned up

File: datafusion/core/src/datasource/file_format/arrow.rs [hunk 5]
```diff
@@ -97,11 +110,191 @@ impl FileFormat for ArrowFormat {
...
+            part_col,
+            self.config.table_paths[0].clone(),
+            "arrow".into(),
+            self.config.single_file_output,
+        );
+
+        let mut file_write_tasks: JoinSet<std::result::Result<usize, DataFusionError>> =
+            JoinSet::new();
+        while let Some((path, mut rx)) = file_stream_rx.recv().await {
+            let shared_buffer = SharedBuffer::new(1048576);
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/arrow.rs:235]: Does this mean that if any record batch takes more than 1MB to write out we'll get an error?

Would it be possible to make this constant and `1024000` below into names constants with comments that explain what they do ?

**Review by alamb — APPROVED** [review]
Thank you for this PR @devinjdangelo  -- I tried it out and it looks quite awesome. 

I also tried it out locally and it works great. 

I left a few suggestions which would be nice to address but I don't think are required to merge. We can do them as a follow on PR as well

```sql
❯ copy (values (1), (2)) to '/tmp/foo.arrow';
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.030 seconds.
```

```shell
datafusion-cli -c "select * from '/tmp/foo.arrow'";
DataFusion CLI v34.0.0
+---------+
| column1 |
+---------+
| 1       |
| 2       |
+---------+
2 rows in set. Query took 0.028 seconds.
```

``shell
$ datafusion-cli -c "select arrow_typeof(column1) from '/tmp/foo.arrow'";
DataFusion CLI v34.0.0
+--------------------------------------+
| arrow_typeof(/tmp/foo.arrow.column1) |
+--------------------------------------+
| Int64                                |
| Int64                                |
+--------------------------------------+
```

👍

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 4 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:211]: Agreed. When we update FileTypeWriterProperties for arrow files, we should also take care to maintain serialization support which @andygrove has been working on.

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 5 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:235]: The initial buffer size is just a size hint for efficiency. It will grow beyond the set value if needed. 

We can definitely make it a named constant, or even make it configurable.

File: datafusion/core/src/datasource/file_format/arrow.rs [hunk 6]
```diff
@@ -97,11 +110,191 @@ impl FileFormat for ArrowFormat {
...
+            let mut arrow_writer = arrow_ipc::writer::FileWriter::try_new(
+                shared_buffer.clone(),
+                &self.get_writer_schema(),
+            )?;
+            let mut object_store_writer = create_writer(
+                FileCompressionType::UNCOMPRESSED,
```
Comment (Dandandan) [inline datafusion/core/src/datasource/file_format/arrow.rs:241]: We probably want to default this to a compressed version (and make it configurable later)? Default of pyarrow is lz4.

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 5 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:235]: This buffer holds serialized bytes  in memory until it is periodically uploaded to an object store. This is similar to how the parquet AsyncArrowWriter for parquet is implemented.

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 4 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:212]: Files #8635 and added comment linking to it

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 6 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:242]: I took a quick look at the arrow_ipc::FileWriter code and it appears that the writer manages compression internally in batches. The referenced line controls whole file compression (like for CSV and JSON). 

Since we are not setting a compression explicitly in DataFusion in this PR, we are inheriting the arrow-rs default compression. I think the arrow-rs default is also lz4, but I am not 100% sure glancing over the code.

Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:249]: Looking more closely, I was wrong. arrow_ipc defaults to uncompressed arrow files. I pushed up a change to default to lz4.

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 4 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/file_format/arrow.rs:214]: Also filed https://github.com/apache/arrow-rs/issues/5236 which would help with #8635 (though is not blocking)

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/src/datasource/file_format/arrow.rs (hunk 5 above)
Comment (alamb) [reply datafusion/core/src/datasource/file_format/arrow.rs:242]: Filed https://github.com/apache/arrow-datafusion/pull/8642 to name the constants
